packages/workflow/src/internals.ts

```diff
@@ -651,10 +676,21 @@ export class Activator implements ActivationHandler {
         this.rejectUpdate(protocolInstanceId, error);
         return;
       }
+
+      const entry = this.updateHandlers.get(name);
+      if (!entry) {
+        return Promise.reject(new IllegalStateError(`No registered update handler for update: ${name}`));
+      }
+      const { unfinishedPolicy } = entry;
+
       const execute = composeInterceptors(this.interceptors.inbound, 'handleUpdate', this.updateNextHandler.bind(this));
       this.acceptUpdate(protocolInstanceId);
+      this.inProgressUpdates.set(updateId, { name, unfinishedPolicy, id: updateId });
       const res = execute(input)
         .then((result) => this.completeUpdate(protocolInstanceId, result))
+        .then(() => {
```

Comment:
Should this be in a `finally()` block?

Reply:
I actually placed it in a `finally()` block initially. However, if the handler is interrupted by cancellation, then we want that to count as an unfinished handler. But if it were in a `finally()` block, a cancellation exception would cause the handler to  be considered completed. There are tests for all these cases. I agree we need to be careful with this logic; do you see a bug with it?

Reply:
> But if it were in a finally() block, a cancellation exception would cause the handler to be considered completed.

But it is completed, why not?

Reply:
Discussed in other channels:

An update handler execution interrupted by a workflow cancellation exception should be considered "finished", since the effect of a cancellation exception is to cancel the update and return an appropriate error to the update caller.

It is not obvious whether a signal handler execution interrupted by a workflow cancellation exception should be considered "finished", but, as a product decision, we don't want to issue unfinished handler warnings on workflow cancellation since the unfinishedness in that case is not something that can be addressed by improving the workflow implementation (it's simply because something cancelled the workflow).

So, the tests need updating here and, yes, the in-progress state transition should be in a `finally()` callback, as it was initially. cc @cretz re .NET

Reply:
Updated implementation and tests: we now issue the warnings neither on cancellation nor on workflow failure.

Reply:
We changed our minds. We are now issuing the warning on cancellation (because the workflow author can avoid leaving unfinished handlers) but not on failure (although there too, the workflow author could avoid it, so it's not clear what the correct choice is).

I've update this PR accordingly, adding tests showing that the warnings _are not_ issued when workflow cancellation propagates to a handler (since then the handler completes before workflow completion) but _are_ issued when the handler shields itself from cancellation and thus does not complete before the main workflow routine terminates due to the cancellation.